File: app/src/main/java/org/puffinbasic/domain/Scope.java
package org.puffinbasic.domain;

import static org.puffinbasic.domain.PuffinBasicSymbolTable.NULL_ID;

import org.puffinbasic.domain.STObjects.STEntry;
import org.puffinbasic.domain.Variable.VariableName;

import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
import it.unimi.dsi.fastutil.objects.Object2IntMap;
import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;

public interface Scope {

    int getCallerInstrId();

    Scope createRuntimeScope(int callerInstrId);

    Scope createChild(int funcId, boolean localScope);

    Scope getChild(int funcId);

    Scope getSearchScope();

    Scope getParent();

    int getIdForVariable(VariableName variableName);

    void putVariable(VariableName variableName, int id);

    boolean containsVariable(VariableName variableName);

    void putEntry(int id, STEntry entry);

    STEntry getEntry(int id);

    STEntry getNullableEntry(int id);

    final class GlobalScope implements Scope {
        private static final int INITIAL_ENTRY_TABLE_SIZE = 1024;
        private final int callerInstrId;
        private final Int2ObjectMap<Scope> funcIdToScope;
        private final Object2IntMap<VariableName> variableNameToEntry;
        // This is an optimization to make entry access fast at runtime.
        //private final ObjectList<STEntry> entryMap;
        private STEntry[] entryMap;

        GlobalScope() {
            this(
                    NULL_ID,
                    new Int2ObjectOpenHashMap<>(),
                    new STEntry[INITIAL_ENTRY_TABLE_SIZE],
                    new Object2IntOpenHashMap<>());
        }

        private GlobalScope(
                int callerInstrId,
                Int2ObjectMap<Scope> funcIdToScope,
                STEntry[] entryMap,
                Object2IntMap<VariableName> variableNameToEntry) {
            this.callerInstrId = callerInstrId;
            this.funcIdToScope = funcIdToScope;
            this.entryMap = entryMap;
            this.variableNameToEntry = variableNameToEntry;
        }

        @Override
        public Scope createRuntimeScope(int callerInstrId) {
            return new GlobalScope(
                    callerInstrId,
                    funcIdToScope,
                    entryMap,
                    variableNameToEntry);
        }

        @Override
        public int getCallerInstrId() {
            return callerInstrId;
        }

        @Override
        public Scope createChild(int funcId, boolean localScope) {
            var child = funcIdToScope.get(funcId);
            if (child == null) {
                child = localScope ? new LocalScope(this) : new ChildScope(this);
                funcIdToScope.put(funcId, child);
            }
            return child;
        }

        @Override
        public Scope getChild(int funcId) {
            return funcIdToScope.get(funcId);
        }

        @Override
        public Scope getParent() {
            return null;
        }

        @Override
        public Scope getSearchScope() {
            return null;
        }

        @Override
        public int getIdForVariable(VariableName variableName) {
            return variableNameToEntry.getOrDefault(variableName, NULL_ID);
        }

        @Override
        public void putVariable(VariableName variableName, int id) {
            variableNameToEntry.put(variableName, id);
        }

        @Override
        public boolean containsVariable(VariableName variableName) {
            return variableNameToEntry.containsKey(variableName);
        }

        private void resize(int index) {
            int newLen = entryMap.length << 1;
            if (newLen < index) {
                do {
                    newLen = newLen << 1;
                } while (newLen < index);
            }
            var newEntryMap = new STEntry[newLen];
            System.arraycopy(entryMap, 0, newEntryMap, 0, entryMap.length);
            entryMap = newEntryMap;
        }

        @Override
        public void putEntry(int id, STEntry entry) {
            int sz = entryMap.length;
            if (id >= sz) {
                resize(id);
            }
            entryMap[id] = entry;
        }

        @Override
        public STEntry getEntry(int id) {
            return entryMap[id];
        }

        @Override
        public STEntry getNullableEntry(int id) {
            if (id >= 0 && id < entryMap.length) {
                return entryMap[id];
            }
            return null;
        }
    }

    final class ChildScope implements Scope {
        private final Scope parent;
        private final int callerInstrId;
        private final Int2ObjectMap<Scope> funcIdToScope;
        private final Int2ObjectMap<STEntry> entryMap;
        private final Object2IntMap<VariableName> variableNameToEntry;

        ChildScope(Scope parent) {
            this(parent,
                    NULL_ID,
                    new Int2ObjectOpenHashMap<>(),
                    new Int2ObjectOpenHashMap<>(),
                    new Object2IntOpenHashMap<>());
        }

        private ChildScope(
                Scope parent,
                int callerInstrId,
                Int2ObjectMap<Scope> funcIdToScope,
                Int2ObjectMap<STEntry> entryMap,
                Object2IntMap<VariableName> variableNameToEntry) {
            this.parent = parent;
            this.callerInstrId = callerInstrId;
            this.funcIdToScope = funcIdToScope;
            this.entryMap = entryMap;
            this.variableNameToEntry = variableNameToEntry;
        }

        @Override
        public Scope createRuntimeScope(int callerInstrId) {
            return new ChildScope(
                    parent,
                    callerInstrId,
                    new Int2ObjectOpenHashMap<>(funcIdToScope),
                    new Int2ObjectOpenHashMap<>(entryMap),
                    new Object2IntOpenHashMap<>(variableNameToEntry)
            );
        }

        @Override
        public int getCallerInstrId() {
            return callerInstrId;
        }

        @Override
        public Scope createChild(int funcId, boolean localScope) {
            var child = funcIdToScope.get(funcId);
            if (child == null) {
                child = new ChildScope(this);
                funcIdToScope.put(funcId, child);
            }
            return child;
        }

        @Override
        public Scope getChild(int funcId) {
            return funcIdToScope.get(funcId);
        }

        @Override
        public Scope getParent() {
            return parent;
        }

        @Override
        public Scope getSearchScope() {
            return parent;
        }

        @Override
        public int getIdForVariable(VariableName variableName) {
            return variableNameToEntry.getOrDefault(variableName, -1);
        }

        @Override
        public void putVariable(VariableName variableName, int id) {
            variableNameToEntry.put(variableName, id);
        }

        @Override
        public boolean containsVariable(VariableName variableName) {
            return variableNameToEntry.containsKey(variableName);
        }

        @Override
        public void putEntry(int id, STEntry entry) {
            entryMap.put(id, entry);
        }

        @Override
        public STEntry getEntry(int id) {
            return entryMap.get(id);
        }

        @Override
        public STEntry getNullableEntry(int id) {
            return entryMap.get(id);
        }
    }

    final class LocalScope implements Scope {
        private final Scope parent;
        private final int callerInstrId;
        private final Int2ObjectMap<Scope> funcIdToScope;
        private final Int2ObjectMap<STEntry> entryMap;
        private final Object2IntMap<VariableName> variableNameToEntry;

        LocalScope(Scope parent) {
            this(parent,
                    NULL_ID,
                    new Int2ObjectOpenHashMap<>(),
                    new Int2ObjectOpenHashMap<>(),
                    new Object2IntOpenHashMap<>());
        }

        private LocalScope(
                Scope parent,
                int callerInstrId,
                Int2ObjectMap<Scope> funcIdToScope,
                Int2ObjectMap<STEntry> entryMap,
                Object2IntMap<VariableName> variableNameToEntry) {
            this.parent = parent;
            this.callerInstrId = callerInstrId;
            this.funcIdToScope = funcIdToScope;
            this.entryMap = entryMap;
            this.variableNameToEntry = variableNameToEntry;
        }

        @Override
        public Scope createRuntimeScope(int callerInstrId) {
            return new LocalScope(
                    parent,
                    callerInstrId,
                    new Int2ObjectOpenHashMap<>(funcIdToScope),
                    new Int2ObjectOpenHashMap<>(entryMap),
                    new Object2IntOpenHashMap<>(variableNameToEntry)
            );
        }

        @Override
        public int getCallerInstrId() {
            return callerInstrId;
        }

        @Override
        public Scope createChild(int funcId, boolean localScope) {
            var child = funcIdToScope.get(funcId);
            if (child == null) {
                child = new ChildScope(this);
                funcIdToScope.put(funcId, child);
            }
            return child;
        }

        @Override
        public Scope getChild(int funcId) {
            return funcIdToScope.get(funcId);
        }

        @Override
        public Scope getParent() {
            return parent;
        }

        @Override
        public Scope getSearchScope() {
            return null;
        }

        @Override
        public int getIdForVariable(VariableName variableName) {
            return variableNameToEntry.getOrDefault(variableName, -1);
        }

        @Override
        public void putVariable(VariableName variableName, int id) {
            variableNameToEntry.put(variableName, id);
        }

        @Override
        public boolean containsVariable(VariableName variableName) {
            return variableNameToEntry.containsKey(variableName);
        }

        @Override
        public void putEntry(int id, STEntry entry) {
            entryMap.put(id, entry);
        }

        @Override
        public STEntry getEntry(int id) {
            return entryMap.get(id);
        }

        @Override
        public STEntry getNullableEntry(int id) {
            return entryMap.get(id);
        }
    }
}

File: app/src/main/java/org/puffinbasic/domain/Scope.kt
package org.puffinbasic.domain

import org.puffinbasic.domain.STObjects.STEntry
import org.puffinbasic.domain.Variable.VariableName

interface Scope {
    val callerInstrId: Int
    fun createRuntimeScope(callerInstrId: Int): Scope
    fun createChild(funcId: Int, localScope: Boolean): Scope
    fun getChild(funcId: Int): Scope
    val searchScope: Scope?
    val parent: Scope?
    fun getIdForVariable(variableName: VariableName): Int
    fun putVariable(variableName: VariableName, id: Int)
    fun containsVariable(variableName: VariableName): Boolean
    fun putEntry(id: Int, entry: STEntry)
    fun getEntry(id: Int): STEntry?
    fun getNullableEntry(id: Int): STEntry?
    class GlobalScope private constructor(
        override val callerInstrId: Int,
        private val funcIdToScope: MutableMap<Int, Scope>,
        // This is an optimization to make entry access fast at runtime.
        //private final ObjectList<STEntry> entryMap;
        private var entryMap: Array<STEntry?>,
        private val variableNameToEntry: MutableMap<VariableName, Int>
    ) : Scope {
        internal constructor() : this(
            PuffinBasicSymbolTable.NULL_ID,
            mutableMapOf(),
            arrayOfNulls<STEntry>(INITIAL_ENTRY_TABLE_SIZE),
            mutableMapOf(),
        )

        override fun createRuntimeScope(callerInstrId: Int): Scope {
            return GlobalScope(
                callerInstrId,
                funcIdToScope,
                entryMap,
                variableNameToEntry
            )
        }

        override fun createChild(funcId: Int, localScope: Boolean): Scope {
            var child = funcIdToScope[funcId]
            if (child == null) {
                child = if (localScope) LocalScope(this) else ChildScope(this)
                funcIdToScope[funcId] = child
            }
            return child
        }

        override fun getChild(funcId: Int): Scope {
            return funcIdToScope[funcId]!!
        }

        override val searchScope: Scope?
            get() = null
        override val parent: Scope?
            get() = null

        override fun getIdForVariable(variableName: VariableName): Int {
            return variableNameToEntry.getOrDefault(
                variableName,
                PuffinBasicSymbolTable.NULL_ID
            )
        }

        override fun putVariable(variableName: VariableName, id: Int) {
            variableNameToEntry[variableName] = id
        }

        override fun containsVariable(variableName: VariableName): Boolean {
            return variableNameToEntry.containsKey(variableName)
        }

        private fun resize(index: Int) {
            var newLen = entryMap.size shl 1
            if (newLen < index) {
                do {
                    newLen = newLen shl 1
                } while (newLen < index)
            }
            val newEntryMap = arrayOfNulls<STEntry>(newLen)
            System.arraycopy(entryMap, 0, newEntryMap, 0, entryMap.size)
            entryMap = newEntryMap
        }

        override fun putEntry(id: Int, entry: STEntry) {
            val sz = entryMap.size
            if (id >= sz) {
                resize(id)
            }
            entryMap[id] = entry
        }

        override fun getEntry(id: Int): STEntry? {
            return entryMap[id]
        }

        override fun getNullableEntry(id: Int): STEntry? {
            return if (id >= 0 && id < entryMap.size) {
                entryMap[id]
            } else null
        }

        companion object {
            private const val INITIAL_ENTRY_TABLE_SIZE = 1024
        }
    }

    class ChildScope private constructor(
        override val parent: Scope,
        override val callerInstrId: Int,
        private val funcIdToScope: MutableMap<Int, Scope>,
        private val entryMap: MutableMap<Int, STEntry>,
        private val variableNameToEntry: MutableMap<VariableName, Int>
    ) : Scope {
        internal constructor(parent: Scope) : this(
            parent,
            PuffinBasicSymbolTable.NULL_ID,
            mutableMapOf(),
            mutableMapOf(),
            mutableMapOf(),
        )

        override fun createRuntimeScope(callerInstrId: Int): Scope {
            return ChildScope(
                parent,
                callerInstrId,
                funcIdToScope.toMutableMap(),
                entryMap.toMutableMap(),
                variableNameToEntry.toMutableMap(),
            )
        }

        override fun createChild(funcId: Int, localScope: Boolean): Scope {
            var child = funcIdToScope[funcId]
            if (child == null) {
                child = ChildScope(this)
                funcIdToScope[funcId] = child
            }
            return child
        }

        override fun getChild(funcId: Int): Scope {
            return funcIdToScope[funcId]!!
        }

        override val searchScope: Scope
            get() = parent

        override fun getIdForVariable(variableName: VariableName): Int {
            return variableNameToEntry.getOrDefault(variableName, -1)
        }

        override fun putVariable(variableName: VariableName, id: Int) {
            variableNameToEntry[variableName] = id
        }

        override fun containsVariable(variableName: VariableName): Boolean {
            return variableNameToEntry.containsKey(variableName)
        }

        override fun putEntry(id: Int, entry: STEntry) {
            entryMap[id] = entry
        }

        override fun getEntry(id: Int): STEntry? {
            return entryMap[id]
        }

        override fun getNullableEntry(id: Int): STEntry? {
            return entryMap[id]
        }
    }

    class LocalScope private constructor(
        override val parent: Scope,
        override val callerInstrId: Int,
        private val funcIdToScope: MutableMap<Int, Scope>,
        private val entryMap: MutableMap<Int, STEntry>,
        private val variableNameToEntry: MutableMap<VariableName, Int>
    ) : Scope {
        internal constructor(parent: Scope) : this(
            parent,
            PuffinBasicSymbolTable.NULL_ID,
            mutableMapOf(),
            mutableMapOf(),
            mutableMapOf(),
        )

        override fun createRuntimeScope(callerInstrId: Int): Scope {
            return LocalScope(
                parent,
                callerInstrId,
                funcIdToScope.toMutableMap(),
                entryMap.toMutableMap(),
                variableNameToEntry.toMutableMap(),
            )
        }

        override fun createChild(funcId: Int, localScope: Boolean): Scope {
            var child = funcIdToScope[funcId]
            if (child == null) {
                child = ChildScope(this)
                funcIdToScope[funcId] = child
            }
            return child
        }

        override fun getChild(funcId: Int): Scope {
            return funcIdToScope[funcId]!!
        }

        override val searchScope: Scope?
            get() = null

        override fun getIdForVariable(variableName: VariableName): Int {
            return variableNameToEntry.getOrDefault(variableName, -1)
        }

        override fun putVariable(variableName: VariableName, id: Int) {
            variableNameToEntry[variableName] = id
        }

        override fun containsVariable(variableName: VariableName): Boolean {
            return variableNameToEntry.containsKey(variableName)
        }

        override fun putEntry(id: Int, entry: STEntry) {
            entryMap[id] = entry
        }

        override fun getEntry(id: Int): STEntry? {
            return entryMap[id]
        }

        override fun getNullableEntry(id: Int): STEntry? {
            return entryMap[id]
        }
    }
}